File: examples/python-agent-simulation/agent_simulation.py
import sys
import os
import time
import json
import urllib.request
import urllib.error

GATRA_PROXY_URL = "http://localhost:8080/v1/action"

def execute_agent_tool_call(token: str, amount: float, currency: str, step_description: str):
    payload = {"amount": amount, "currency": currency}
    data = json.dumps(payload).encode("utf-8")

    print(f"\n🤖 [Agent Decision] {step_description}")
    print(f"   Payload -> Amount: ${amount:.2f} {currency}")

    headers = {
        "X-Capability-Token": token,
        "Content-Type": "application/json",
    }

    req = urllib.request.Request(GATRA_PROXY_URL, data=data, headers=headers, method="POST")
    start_time = time.perf_counter()

    try:
        with urllib.request.urlopen(req) as res:
            latency_ms = (time.perf_counter() - start_time) * 1000
            resp_data = json.loads(res.read().decode("utf-8"))
            print(f"   🟢 [GATRA APPROVED] HTTP {res.status} ({latency_ms:.2f}ms)")
            print(f"      Downstream Response: {resp_data}")
    except urllib.error.HTTPError as err:
        latency_ms = (time.perf_counter() - start_time) * 1000
        try:
            resp_data = json.loads(err.read().decode("utf-8"))
        except Exception:
            resp_data = err.reason
        print(f"   🔴 [GATRA INTERCEPTED & BLOCKED] HTTP {err.code} ({latency_ms:.2f}ms)")
        print(f"      Security Gate: {resp_data}")
    except Exception as err:
        print(f"   ❌ Request failed: {err}")

def main():
    token = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("GATRA_TOKEN")

    if not token:
        print("❌ Missing capability token!")
        print("   Usage: python agent_simulation.py <YOUR_ACTIVE_TOKEN>")
        sys.exit(1)

    print("==================================================")
    print("🤖 Simulating Live LLM Agent Execution Session (Python)")
    print("==================================================")
    print(f"🎟️ Active Capability Token:\n   {token[:45]}...")

    # Scenario 1: Approved valid tool invocation
    execute_agent_tool_call(token, 25.0, "USD", "Executing valid tool action $25.00 USD")

    # Scenario 2: CEL Policy Rejection (Forbidden currency)
    execute_agent_tool_call(token, 15.0, "EUR", "Attempting forbidden currency €15.00 EUR")

    # Scenario 3: Single-call limit breach ($60 > $50 max per call)
    execute_agent_tool_call(token, 60.0, "USD", "Attempting limit breach $60.00 USD (> $50 max)")

    # Scenario 4: Cumulative trajectory boundary breach ($40 x 5 > $200 limit)
    print("\n--------------------------------------------------")
    print("🔄 Testing Stateful Cumulative Trajectory Boundary")
    print("--------------------------------------------------")

    for i in range(1, 6):
        execute_agent_tool_call(token, 40.0, "USD", f"Cumulative Action #{i} ($40.00 USD)")

if __name__ == "__main__":
    main()

File: examples/node-agent-simulation/agent_simulation.js
const GATRA_PROXY_URL = "http://localhost:8080/v1/refund";

async function executeAgentToolCall(token, amount, currency, stepDescription) {
  const payload = { amount, currency };

  console.log(`\n🤖 [Agent Decision] ${stepDescription}`);
  console.log(`   Payload -> Amount: $${amount.toFixed(2)} ${currency}`);

  const startTime = performance.now();

  try {
    const res = await fetch(GATRA_PROXY_URL, {
      method: "POST",
      headers: {
        "X-Capability-Token": token,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });

    const latencyMs = (performance.now() - startTime).toFixed(2);
    const data = await res.json();

    if (res.ok) {
      console.log(`   🟢 [GATRA APPROVED] HTTP ${res.status} (${latencyMs}ms)`);
      console.log(`      Downstream Response:`, data);
    } else {
      console.log(`   🔴 [GATRA INTERCEPTED & BLOCKED] HTTP ${res.status} (${latencyMs}ms)`);
      console.log(`      Security Gate:`, data);
    }
  } catch (err) {
    console.error(`   ❌ Request failed:`, err);
  }
}

async function main() {
  const token = process.argv[2] || process.env.GATRA_TOKEN;

  if (!token) {
    console.error("❌ Missing capability token!");
    console.error("   Usage: node agent_simulation.js <YOUR_ACTIVE_TOKEN>");
    process.exit(1);
  }

  console.log("==================================================");
  console.log("🤖 Simulating Live LLM Agent Execution Session (Node.js)");
  console.log("==================================================");
  console.log(`🎟️ Active Capability Token:\n   ${token.substring(0, 45)}...`);

  // Scenario 1: Approved valid tool invocation
  await executeAgentToolCall(token, 25.0, "USD", "Refunding $25.00 USD for customer ticket #1042");

  // Scenario 2: CEL Policy Rejection (Forbidden currency)
  await executeAgentToolCall(token, 15.0, "EUR", "Refunding €15.00 EUR for European user ticket #1043");

  // Scenario 3: Single-call limit breach ($60 > $50 max per call)
  await executeAgentToolCall(token, 60.0, "USD", "Attempting high-value refund of $60.00 USD");

  // Scenario 4: Cumulative trajectory boundary breach ($40 x 5 > $200 limit)
  console.log("\n--------------------------------------------------");
  console.log("🔄 Testing Stateful Cumulative Trajectory Boundary");
  console.log("--------------------------------------------------");

  for (let i = 1; i <= 5; i++) {
    await executeAgentToolCall(token, 40.0, "USD", `Cumulative Refund #${i} ($40.00 USD)`);
  }
}

main();